taskchampion/taskchampion/src/server/cloud/gcp.rs

```diff
@@ -0,0 +1,392 @@
+                    ..Default::default()
+                },
+                &objects::download::Range::default(),
+            ))?;
+            if Some(data) != existing_value {
+                return Ok(false);
+            }
+        }
+
+        // Finally, put the new value with a condition that the generation hasn't changed.
+        let upload_type = objects::upload::UploadType::Simple(objects::upload::Media::new(name));
+        let upload_res = self.rt.block_on(self.client.upload_object(
+            &objects::upload::UploadObjectRequest {
+                bucket: self.bucket.clone(),
+                if_generation_match: Some(generation),
+                ..Default::default()
+            },
+            new_value.to_vec(),
+            &upload_type,
+        ));
+        if is_http_error(412, &upload_res) {
+            // A 412 indicates the precondition was not satisfied: the given generation
+            // is no longer the latest.
+            Ok(false)
+        } else {
+            upload_res?;
+            Ok(true)
+        }
+    }
+}
+
+/// An Iterator returning names of objects from `list_objects`.
+///
+/// This handles response pagination by fetching one page at a time.
+struct ObjectIterator<'a> {
+    service: &'a mut GcpService,
+    prefix: String,
+    last_response: Option<objects::list::ListObjectsResponse>,
+    next_index: usize,
+}
+
+impl<'a> ObjectIterator<'a> {
+    fn fetch_batch(&mut self) -> Result<()> {
+        let mut page_token = None;
+        if let Some(ref resp) = self.last_response {
+            page_token = resp.next_page_token.clone();
+        }
+        self.last_response = Some(self.service.rt.block_on(self.service.client.list_objects(
+            &objects::list::ListObjectsRequest {
+                bucket: self.service.bucket.clone(),
+                prefix: Some(self.prefix.clone()),
+                page_token,
+                #[cfg(test)] // For testing, use a small page size.
+                max_results: Some(6),
+                ..Default::default()
+            },
+        ))?);
+        self.next_index = 0;
+        Ok(())
+    }
+}
+
+impl<'a> Iterator for ObjectIterator<'a> {
+    type Item = Result<ObjectInfo>;
+    fn next(&mut self) -> Option<Self::Item> {
+        // If the iterator is just starting, fetch the first response.
+        if self.last_response.is_none() {
+            if let Err(e) = self.fetch_batch() {
+                return Some(Err(e));
+            }
+        }
+        if let Some(ref result) = self.last_response {
+            if let Some(ref items) = result.items {
+                if self.next_index < items.len() {
+                    // Return a result from the existing response.
+                    let obj = &items[self.next_index];
+                    self.next_index += 1;
+                    let creation = obj.time_created.map(|t| t.unix_timestamp()).unwrap_or(0);
+                    let creation: u64 = creation.try_into().unwrap_or(0);
```

Comment:
It's not clear from the docs when this would be None (the first `unwrap_or`). The second `unwrap_or` is for when the timestamp doesn't fit into a `u64`: either in 300 billion years or more likely before 1970.

In both cases, the object will appear to have been created in 1970, which may lead to it being cleaned up if it is included in a snapshot -- not an especially problematic situation. I think that's better than failing, especially in the unknown conditions where the value is None.

I'll add a comment.